odoo/models.py

```diff
@@ -1948,41 +1948,27 @@ def _read_group(self, domain, groupby=(), aggregates=(), having=(), offset=0, li
             return []
 
         query = self._search(domain)
+        query.limit = limit
+        query.offset = offset
 
         groupby_terms: dict[str, SQL] = {
             spec: self._read_group_groupby(spec, query)
             for spec in groupby
         }
+        if groupby_terms:
+            query.groupby = SQL(", ").join(groupby_terms.values())
+            query.having = self._read_group_having(having, query)
+            # _read_group_orderby will extend query.groupby for orderby
+            query.order = self._read_group_orderby(order, groupby_terms, query)
+
         select_terms: list[SQL] = [
             self._read_group_select(spec, query)
             for spec in aggregates
         ]
-        sql_having = self._read_group_having(having, query)
-        sql_order, sql_extra_groupby = self._read_group_orderby(order, groupby_terms, query)
-
         groupby_terms = list(groupby_terms.values())
 
-        query_parts = [
-            SQL("SELECT %s", SQL(", ").join(groupby_terms + select_terms)),
-            SQL("FROM %s", query.from_clause),
-        ]
-        if query.where_clause:
-            query_parts.append(SQL("WHERE %s", query.where_clause))
-        if groupby_terms:
-            if sql_extra_groupby:
-                groupby_terms.append(sql_extra_groupby)
-            query_parts.append(SQL("GROUP BY %s", SQL(", ").join(groupby_terms)))
-        if sql_having:
-            query_parts.append(SQL("HAVING %s", sql_having))
-        if sql_order:
-            query_parts.append(SQL("ORDER BY %s", sql_order))
-        if limit:
-            query_parts.append(SQL("LIMIT %s", limit))
-        if offset:
-            query_parts.append(SQL("OFFSET %s", offset))
-
         # row_values: [(a1, b1, c1), (a2, b2, c2), ...]
-        row_values = self.env.execute_query(SQL("\n").join(query_parts))
+        row_values = self.env.execute_query(query.select(*groupby_terms, *select_terms))
```

Comment:
Maybe remove the (type-changing) assignment
```py
groupby_terms = list(groupby_terms.values())
```
above and use
```suggestion
        row_values = self.env.execute_query(query.select(*groupby_terms.values(), *select_terms))
```